include/RAJA/internal/foldl.hpp

```diff
@@ -0,0 +1,132 @@
+/*!
+ ******************************************************************************
+ *
+ * \file
+ *
+ * \brief   Header file with support for pre-C++14 compilers.
+ *
+ ******************************************************************************
+ */
+
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2016-20, Lawrence Livermore National Security, LLC
+// and RAJA project contributors. See the RAJA/COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#ifndef RAJA_foldl_HPP
+#define RAJA_foldl_HPP
+
+#include "RAJA/config.hpp"
+
+#include <cstdint>
+#include <functional>
+#include <iostream>
+#include <type_traits>
+#include <utility>
+
+#include "camp/camp.hpp"
+
+#include "RAJA/util/macros.hpp"
+
+
+namespace RAJA
+{
+
+// Basics, using c++14 semantics in a c++11 compatible way, credit to libc++
+
+// Forward
+namespace detail
+{
+// FoldL
+template <typename Op, typename... Rest>
+struct foldl_impl;
```

Comment:
I'll double check, but to your point I think this is in camp already as `accumulate`, if there's a notable difference I'll hoist foldl as well, or maybe just have an alias for it.  I did some of the same refactoring in a branch working toward a camp update, would be good to get this in so I can make sure my changes play nice with what you set up.

Reply:
@trws  RAJA::foldl operates on a list of values, whereas camp::accumulate appears to operate on a type list.   So I think they are both needed.  I think it makes sense to move foldl into camp at some point.